components/cam/src/physics/cam/phys_grid.F90

```diff
@@ -4540,13 +4704,81 @@ subroutine create_chunks(opt, chunks_per_thread)
 !
 !  Allocate other work space
 !
-   allocate( nvsmpthreads(0:nvsmp-1) )
-   allocate( nvsmpchunks (0:nvsmp-1) )
-   allocate( maxcol_chk  (0:nvsmp-1) )
-   allocate( maxcol_chks (0:nvsmp-1) )
-   allocate( cid_offset  (0:nvsmp-1) )
-   allocate( local_cid   (0:nvsmp-1) )
-   allocate( cols_d    (1:maxblksiz) )
+   allocate( nvsmpthreads  (0:nvsmp-1) )
+   allocate( nvsmpchunks   (0:nvsmp-1) )
+   allocate( pcols_vsmp    (0:nvsmp-1) )
+   allocate( maxcol_chk    (0:nvsmp-1) )
+   allocate( maxcol_chks   (0:nvsmp-1) )
+   allocate( maxblksiz_vsmp(0:nvsmp-1) )
+   allocate( cid_offset    (0:nvsmp-1) )
+   allocate( local_cid     (0:nvsmp-1) )
+   allocate( cols_d      (1:maxblksiz) )
+
+!
+! Calculate number of threads available in each virtual SMP node. 
+!
+   nvsmpthreads(:) = 0
+   do p=0,npes-1
+      smp = proc_vsmp_map(p)
+      nvsmpthreads(smp) = nvsmpthreads(smp) + npthreads(p)
+   enddo
+
+!
+! Calculate maximum block size for each virtual SMP
+!
+   maxblksiz_vsmp(:) = 0
+   do j=firstblock,lastblock
+      p = get_block_owner_d(j)
+      smp = proc_vsmp_map(p)
+      blksiz = get_block_gcol_cnt_d(j)
+      if (blksiz > maxblksiz_vsmp(smp)) maxblksiz_vsmp(smp) = blksiz
+   enddo
+
+!
+! Calculate pcols for chunks in each virtual SMP
+!
+#ifdef PPCOLS
+! use compile-time value
+   pcols_vsmp(:) = pcols
+#else
+   if (cols_per_chnk > 0) then
+! use runtime value
+      pcols_vsmp(:) = cols_per_chnk
+   else
+      do smp=0,nvsmp-1
+! pick pcols to minimize wasted space and the number of chunks per thread
+         pcols_vsmp(smp) = &
+            ceiling(real(nvsmpcolumns(smp),r8)/real(nvsmpthreads(smp)*chnks_per_thrd,r8))
+         if (opt == 5) then
+! pcols should not be (much) larger than the maximum block size
+            if (maxblksiz_vsmp(smp) < pcols_vsmp(smp)) pcols_vsmp(smp) = maxblksiz_vsmp(smp)
+         endif
+! then increase (if necessary) to be a multiple of cols_per_chnk_mult
+         if (cols_per_chnk_mult > 1) then
+            pcols_vsmp(smp) = &
+               cols_per_chnk_mult*ceiling(real(pcols_vsmp(smp),r8)/real(cols_per_chnk_mult,r8))
+         endif
+         if (cols_per_chnk_max > 0) then
+! if calculated pcols is too large, recalculate with more chunks per thread
+            tmp_chnks_per_thrd = chnks_per_thrd
+            do while (pcols_vsmp(smp) > cols_per_chnk_max)
+               tmp_chnks_per_thrd = tmp_chnks_per_thrd + 1
+               pcols_vsmp(smp) = &
+                  ceiling(real(nvsmpcolumns(smp),r8)/real(nvsmpthreads(smp)*tmp_chnks_per_thrd,r8))
+               if (opt == 5) then
+! pcols should not be (much) larger than the maximum block size
```

Comment:
To match the style above I think these single comment lines should be indented. Personally, I find the left-justified comment make the logic/loop structure harder to parse visually, so it becomes a distraction.

Reply:
There is a mix of styles in this routine, but, you are right, I picked the less used style. I'll rework this, at least cleaning up the comments in the code that I just added. Thanks.

Reply:
@whannah1 , I made a sweep through the routine create_chunks (which the above code is from) to use a consistent style for all comments. I adopted what I think that you prefer. This has been pushed to the branch.

Reply:
I would actually prefer that the "bang" character is also indented. The non-indented style seems to harken back to the F77 days.

Reply:
Done. Hopefully close enough that further tweaking can wait until this part of the code is refactored by introducing more modularity.

Reply:
@worleyph Thanks for doing that!